File: sdks/typescript/src/polling.ts
import type { components } from "./generated/schema";
import { LiteGenPollingTimeoutError } from "./errors";

type VideoResponse = components["schemas"]["VideoGenerationResponse"];
type Model3dResponse = components["schemas"]["Model3dGenerationResponse"];

/** Any pollable job response. */
type WithStatus = { status: string };

export interface WaitForCompletionOptions {
  /** Milliseconds between polls. Default 2000. */
  intervalMs?: number;
  /** Total timeout in milliseconds. Default 5 minutes. */
  timeoutMs?: number;
  /** Optional AbortSignal to cancel polling. */
  signal?: AbortSignal;
}

const TERMINAL_STATUSES = new Set<string>(["completed", "failed", "cancelled"]);

/**
 * What a timeout on this family calls itself. Exported so the 3D wrappers that
 * go through the generic `waitForJob` report the same word as `poll3d`.
 */
export const MODEL3D_KIND = "3D generation";

/**
 * Poll any async job, yielding every status update — including the terminal one
 * — as an async iterable. `progress` is LiteGen's unified 0–100 value across
 * every provider and modality, so this loop behaves identically for video and
 * 3D — providers that don't report fine-grained progress simply step toward
 * 100. Iteration stops after the first terminal status (`completed` / `failed`
 * / `cancelled`), which is the final value yielded.
 *
 * `pollVideo` / `poll3d` are thin instantiations of this generator, not
 * separate implementations — a new async media family gets polling for free
 * by wrapping this with its own response type.
 */
export async function* pollJob<T extends WithStatus>(
  id: string,
  getStatus: (id: string) => Promise<T>,
  opts: WaitForCompletionOptions = {},
  kind = "job",
): AsyncGenerator<T, T, void> {
  const intervalMs = opts.intervalMs ?? 2000;
  const timeoutMs = opts.timeoutMs ?? 5 * 60_000;
  const deadline = Date.now() + timeoutMs;

  let last: T | undefined;
  while (true) {
    if (opts.signal?.aborted) {
      throw new DOMException("Polling aborted", "AbortError");
    }
    if (Date.now() > deadline) {
      throw new LiteGenPollingTimeoutError(id, last?.status, kind);
    }
    last = await getStatus(id);
    yield last;
    if (TERMINAL_STATUSES.has(last.status)) {
      return last;
    }
    await sleep(intervalMs, opts.signal);
  }
}

/** Resolve once any job reaches a terminal status, returning the final state. */
export async function waitForJob<T extends WithStatus>(
  id: string,
  getStatus: (id: string) => Promise<T>,
  opts: WaitForCompletionOptions = {},
  kind = "job",
): Promise<T> {
  let last: T | undefined;
  for await (const update of pollJob(id, getStatus, opts, kind)) {
    last = update;
  }
  // pollJob always yields at least once before completing.
  return last as T;
}

/**
 * Poll a video job. `for await (const update of client.videos.poll(job.id))`.
 */
export const pollVideo = (
  id: string,
  getStatus: (id: string) => Promise<VideoResponse>,
  opts?: WaitForCompletionOptions,
): AsyncGenerator<VideoResponse, VideoResponse, void> =>
  pollJob<VideoResponse>(id, getStatus, opts, "video");

/**
 * Poll a 3D generation job. `for await (const update of client.models3d.poll(job.id))`.
 */
export const poll3d = (
  id: string,
  getStatus: (id: string) => Promise<Model3dResponse>,
  opts?: WaitForCompletionOptions,
): AsyncGenerator<Model3dResponse, Model3dResponse, void> =>
  pollJob<Model3dResponse>(id, getStatus, opts, MODEL3D_KIND);

/**
 * Back-compat wrapper. Deliberately concrete rather than a bare alias for
 * `waitForJob`: exporting the generic directly makes TypeScript infer `T` from
 * the caller's `getStatus`, which silently changes inference at existing call
 * sites that relied on the concrete `VideoResponse` signature.
 */
export const waitForCompletion = (
  id: string,
  getStatus: (id: string) => Promise<VideoResponse>,
  opts?: WaitForCompletionOptions,
): Promise<VideoResponse> => waitForJob<VideoResponse>(id, getStatus, opts, "video");

function sleep(ms: number, signal?: AbortSignal): Promise<void> {
  return new Promise((resolve, reject) => {
    const t = setTimeout(() => resolve(), ms);
    signal?.addEventListener(
      "abort",
      () => {
        clearTimeout(t);
        reject(new DOMException("Polling aborted", "AbortError"));
      },
      { once: true },
    );
  });
}

File: sdks/python/litegen/polling.py
"""Polling helpers for async video generation."""
from __future__ import annotations

import asyncio
import time
from typing import Any, AsyncIterator, Awaitable, Callable, Iterator

from .errors import LiteGenPollingTimeoutError

_TERMINAL = {"completed", "failed", "cancelled"}


def poll_video(
    video_id: str,
    get_status: Callable[[str], dict[str, Any]],
    *,
    interval: float = 2.0,
    timeout: float = 300.0,
) -> Iterator[dict[str, Any]]:
    """Yield each status update for a video job, including the terminal one.

    Drive it with a ``for`` loop::

        for update in client.videos.poll(job["id"]):
            print(update["status"], update["progress"])

    ``progress`` is LiteGen's unified 0-100 value; providers that don't report
    fine-grained progress simply step toward 100, so this loop behaves the same
    across every provider. Iteration stops after the first terminal status
    (``completed`` / ``failed`` / ``cancelled``).
    """
    deadline = time.monotonic() + timeout
    last: dict[str, Any] | None = None
    while True:
        if time.monotonic() > deadline:
            raise LiteGenPollingTimeoutError(
                video_id, _status_value(last) if last is not None else None
            )
        last = get_status(video_id)
        yield last
        if _status_value(last) in _TERMINAL:
            return
        time.sleep(interval)


async def apoll_video(
    video_id: str,
    get_status: Callable[[str], Awaitable[dict[str, Any]]],
    *,
    interval: float = 2.0,
    timeout: float = 300.0,
) -> AsyncIterator[dict[str, Any]]:
    """Async version of :func:`poll_video` — drive with ``async for``."""
    deadline = time.monotonic() + timeout
    last: dict[str, Any] | None = None
    while True:
        if time.monotonic() > deadline:
            raise LiteGenPollingTimeoutError(
                video_id, _status_value(last) if last is not None else None
            )
        last = await get_status(video_id)
        yield last
        if _status_value(last) in _TERMINAL:
            return
        await asyncio.sleep(interval)


def wait_for_video_completion(
    video_id: str,
    get_status: Callable[[str], dict[str, Any]],
    *,
    interval: float = 2.0,
    timeout: float = 300.0,
) -> dict[str, Any]:
    """Synchronously poll until the video reaches a terminal status."""
    last: dict[str, Any] = {}
    for last in poll_video(video_id, get_status, interval=interval, timeout=timeout):
        pass
    return last


async def async_wait_for_video_completion(
    video_id: str,
    get_status: Callable[[str], Awaitable[dict[str, Any]]],
    *,
    interval: float = 2.0,
    timeout: float = 300.0,
) -> dict[str, Any]:
    """Async version of `wait_for_video_completion`."""
    last: dict[str, Any] = {}
    async for last in apoll_video(
        video_id, get_status, interval=interval, timeout=timeout
    ):
        pass
    return last


def _status_value(payload: dict[str, Any] | None) -> str | None:
    if payload is None:
        return None
    status = payload.get("status")
    return getattr(status, "value", status if isinstance(status, str) else None)
